### PR TITLE
[CIR][NFC] Add scaffolding for the CIR dialect and CIROps.td

### PR DESCRIPTION
**Pull request by lanza**
This adds no real content, it just incrementally adds some scaffolding
necessary to enable a future patch to just add our first few ops.

Test Plan:
```
$ cmake -Sllvm -Bbuild -DCLANG_ENABLE_CIR=1 \
  -DLLVM_ENABLE_PROJECTS='clang;mlir' \
  -DCMAKE_BUILD_TYPE=Release -GNinja
$ ninja -C build check-clang
$ ninja -C build MLIRCIROpsIncGen
$ ninja -C build MLIRCIR
```

### CODE REVIEW

**Review by bcardosolopes — COMMENTED** [review]
Maybe add the "comment header" for `CIRDialect.cpp` so that we don't land an empty file?

File: clang/include/clang/CIR/Dialect/IR/CIRDialect.td
Comment (bcardosolopes) [inline clang/include/clang/CIR/Dialect/IR/CIRDialect.td:1]: `CIRTypes.td` ->  `CIRDialect.td`

**Review by bcardosolopes — COMMENTED** [review]
`CIR/Dialect/IR/CIRDialect.h` should get a comment header too, and could likely be already included by `CIRDialect.cpp`.

File: clang/include/clang/CIR/Dialect/IR/CIRDialect.td
Comment (tschuett) [inline clang/include/clang/CIR/Dialect/IR/CIRDialect.td:14]: Thanks! Should the include guards start with clang?

File: clang/include/clang/CIR/Dialect/IR/CIRDialect.td
Comment (AaronBallman) [inline clang/include/clang/CIR/Dialect/IR/CIRDialect.td:28]: Should these follow the usual style guide for naming of member variables (start with a capital letter)?

File: clang/include/clang/CIR/Dialect/IR/CIROps.td
Comment (AaronBallman) [inline clang/include/clang/CIR/Dialect/IR/CIROps.td:14]: Prefix with `CLANG_` as well?

**Review by AaronBallman — COMMENTED** [review]
Generally LGTM as well, but added the CMake code owners for their opinions on those bits.

File: clang/include/clang/CIR/Dialect/IR/CIRDialect.td
Comment (bcardosolopes) [reply clang/include/clang/CIR/Dialect/IR/CIRDialect.td:28]: We are following MLIR style guide here and these tablegen ones need to be like this in order for certain c++ code to be generated.

This comment reminds me of something we need to discuss for the rest of CIR: MLIR and clang differs on the coding style, which means that CIRGen code has variables starting with capital letter calling into code using lowecase. So far we've been using a mix: in the dialect source we keep MLIR style, and in CIRGen we keep the clang style. 

We don't have a strong opinion here, we rather apply whatever the community/maintainers care and move on with upstreaming. Thoughts? (cc: @joker-eph)

Comment (joker-eph) [reply clang/include/clang/CIR/Dialect/IR/CIRDialect.td:28]: If the comment is on the TableGen code: you don't have a choice here because you're implement the ODS API. These aren't just "data member", they are actually "magic" names used by the tablegen backend.

For the C++ code a few lines below, function like `printType()` are override of virtual methods defined in MLIR.

File: clang/include/clang/CIR/Dialect/IR/CIRDialect.td
Comment (joker-eph) [inline clang/include/clang/CIR/Dialect/IR/CIRDialect.td:42]: Nit: you are defining this dialect in the ::mlir::cir namespace, so you can remove the ::mlir:: prefix everywhere here.

File: clang/include/clang/CIR/Dialect/IR/CIROps.td
Comment (joker-eph) [reply clang/include/clang/CIR/Dialect/IR/CIROps.td:14]: It not "as well", I believe it should be "instead of `MLIR_`" here?
Also `IR` is missing, the path mangles into `CLANG_CIR_DIALECT_IR_CIR_OPS_TD` I think?

File: clang/include/clang/CIR/Dialect/IR/CMakeLists.txt
Comment (joker-eph) [inline clang/include/clang/CIR/Dialect/IR/CMakeLists.txt:15]: You're not generating the dialect doc?

Comment (joker-eph) [reply clang/include/clang/CIR/Dialect/IR/CMakeLists.txt:15]: Also: can these lines be abstracted by just calling the `add_mlir_dialect` macro?

File: clang/include/clang/CIR/Dialect/IR/CIRDialect.td
Comment (AaronBallman) [reply clang/include/clang/CIR/Dialect/IR/CIRDialect.td:28]: Okay. so this is another case of MLIR not following the usual community policies, so this is outside of the scope of these patches and that's fine by me. Thanks for the explanation!

File: clang/include/clang/CIR/Dialect/IR/CIROps.td
Comment (AaronBallman) [reply clang/include/clang/CIR/Dialect/IR/CIROps.td:14]: Sorry I was unclear, I meant "as well" in reference to an earlier suggestion: https://github.com/llvm/llvm-project/pull/86080#discussion_r1533453807

File: clang/include/clang/CIR/Dialect/IR/CIRDialect.td
Comment (joker-eph) [reply clang/include/clang/CIR/Dialect/IR/CIRDialect.td:28]: Yes, in the early months of MLIR, there was a big discussion on llvm-dev@ about changing the naming convention to start with lower-cases. And it seemed to us that it converged and so on our side we anticipated and updated all the codebase to use the "new" convention. However this didn't carry forward, I don't remember why...
(The plan was even documented: https://llvm.org/docs/Proposals/VariableNames.html )

Comment (joker-eph) [reply clang/include/clang/CIR/Dialect/IR/CIRDialect.td:28]: Thread: https://groups.google.com/g/llvm-dev/c/i7TLvzd1OAw/m/cJg8sGnMDAAJ

File: clang/include/clang/CIR/CMakeLists.txt
Comment (petrhosek) [inline clang/include/clang/CIR/CMakeLists.txt:1]: This variable seems unused and has the same value as `MLIR_INCLUDE_DIR` below, what's the use case for this variable?

File: clang/include/clang/CIR/CMakeLists.txt
Comment (petrhosek) [inline clang/include/clang/CIR/CMakeLists.txt:5]: Why do these need to be system include directories? Rather than adding these to global list of include paths, could we create an interface library and use it as an explicit dependency only for targets that need it?

Comment (lanza) [reply clang/include/clang/CIR/CMakeLists.txt:5]: `SYSTEM` wasn't necessary! Removing in update.

File: clang/include/clang/CIR/Dialect/IR/CMakeLists.txt
Comment (lanza) [reply clang/include/clang/CIR/Dialect/IR/CMakeLists.txt:15]: > You're not generating the dialect doc?

These series of diffs are made by taking the tip of clangir and removing what we don't need to get a reviewable chunk of code up for a single diff. I guess while chunking this I just excluded it. It's in upstream clangir. (See https://github.com/llvm/clangir/blob/main/clang/include/clang/CIR/Dialect/IR/CMakeLists.txt#L17). 

> Also: can these lines be abstracted by just calling the add_mlir_dialect macro?

For reasons described at the top of the file that did not work. Perhaps that suggests a refactor of `add_mlir_dialect`, but don't intend to address that here.

**Review by bcardosolopes — APPROVED** [review]

File: clang/include/clang/CIR/Dialect/IR/CIRDialect.td
Comment (AaronBallman) [inline clang/include/clang/CIR/Dialect/IR/CIRDialect.td:34]: Because this is new code, the parameter names should match the usual coding style for naming (start with a capital letter) for consistency with the rest of the code base.

Where do the definitions for these come from?

Comment (joker-eph) [reply clang/include/clang/CIR/Dialect/IR/CIRDialect.td:34]: This will be added to generated class which have lower-case starting APIs: consistency within the class seems to prevail to me.
You should really expect that all the MLIR code will follow this convention, because it'll be intertwined closely with the MLIR framework.

Comment (AaronBallman) [reply clang/include/clang/CIR/Dialect/IR/CIRDialect.td:34]: > This will be added to generated class which have lower-case starting APIs: consistency within the class seems to prevail to me.

Ah, then yeah, I agree. Thank you!

> You should really expect that all the MLIR code will follow this convention, because it'll be intertwined closely with the MLIR framework.

MLIR isn't old enough to really get "grandfathered" in with their own style, so that's pretty surprising. That said, I *fully support* letting projects have project-specific policies including coding style, so I see no reason for MLIR to change either. Interfacing closely with MLIR does make this a bit awkward however, because it's also intertwined with the Clang code base. On balance, I think it's fine to follow the MLIR convention because I think there will be more surface area in common between CIR + MLIR than CIR + Clang (and Clang has plenty of inconsistency with coding style already so this is not really novel).

tl;dr: nothing needs to change here.

File: clang/include/clang/CIR/Dialect/IR/CIRDialect.h
Comment (erichkeane) [reply clang/include/clang/CIR/Dialect/IR/CIRDialect.h:1]: I see this is irrelevant now, but noticed I had a pair of pending comments and wanted the other one.

**Review by AaronBallman — APPROVED** [review]
LGTM, though it would be nice if @petrhosek or @Ericson2314 could validate the cmake changes (they look sensible to me, but I never know if there's a better way to do things in CMake).

**Comment by lanza** [comment]
> LGTM, though it would be nice if @petrhosek or @Ericson2314 could validate the cmake changes (they look sensible to me, but I never know if there's a better way to do things in CMake).

We already did that above. Petr left a comment and it was fixed :p

**Comment by AaronBallman** [comment]
> > LGTM, though it would be nice if @petrhosek or @Ericson2314 could validate the cmake changes (they look sensible to me, but I never know if there's a better way to do things in CMake).
> 
> We already did that above. Petr left a comment and it was fixed :p

Ah, I missed that entirely because the conversations were resolved. Phew, then yeah, LGTM!

**Review by erichkeane — APPROVED** [review]

File: clang/lib/CIR/CMakeLists.txt
Comment (Ericson2314) [inline clang/lib/CIR/CMakeLists.txt:2]: I think something like `target_include_directories` is generally preferred, because it is less of a sledgehammer effecting many targets at once.

File: clang/include/clang/CIR/Dialect/IR/CIRDialect.td
Comment (joker-eph) [reply clang/include/clang/CIR/Dialect/IR/CIRDialect.td:34]: I think I mentioned it elsewhere, but worth keeping in mind: MLIR didn't intend to diverge from LLVM, it started about at the time where there was this plan to adopt a new style for LLVM: https://llvm.org/docs/Proposals/VariableNames.html
So we thought we were just adopting the "new coding style" to limit churn, and unfortunately this didn't pan out for the whole codebase (I think lldb and lld are also using this style now)

File: clang/include/clang/CIR/CMakeLists.txt
Comment (lanza) [reply clang/include/clang/CIR/CMakeLists.txt:2]: @Ericson2314 any protest against marking that as just "not yet supported?"

e.g. 
```
if (CLANG_ENABLE_CIR && CLANG_STANDALONE_BUILD) 
  message(ERROR "ClangIR is not yet supported in the standalone build")
endif()
```

**Pull request merged**
